File: src/stplay.cpp
#include <stdlib.h>
#include <time.h>
#include "stplay.h"


/************************************************************
 *	@brief		Constructor
 *	@retval		Nothing
 ************************************************************/
CStPlay::CStPlay(CStView* pView)
	: m_CurrentBlock(&m_Board),
	m_NextBlock(&m_Board)
{
    // View
    m_pView = pView;
}

/************************************************************
 *	@brief		Move left
 *	@retval		Nothing
 ************************************************************/
void CStPlay::MoveLeft()
{
    // Move left
	m_CurrentBlock.MoveLeft();
    // Update view
    m_pView->UpdateView();
}

/************************************************************
 *	@brief		Move right
 *	@retval		Nothing
 ************************************************************/
void CStPlay::MoveRight()
{
    // Move right
	m_CurrentBlock.MoveRight();
    // Update view
    m_pView->UpdateView();
}

/************************************************************
 *	@brief		Move down
 *	@retval		Nothing
 ************************************************************/
void CStPlay::MoveDown()
{
    // Move down
    if (!m_CurrentBlock.MoveDown())
    {
		// Set block to board
		SetBlockToBoard();
		// Clear complete line
		ClearCompleteLine();
		// Change block
		ChangeBlock();
    }

	// Add score
	m_Score.AddScore(10);

	// Set high score
	if (m_HighScore.GetScore() < m_Score.GetScore())
		m_HighScore.SetScore(m_Score.GetScore());

    // Update view
    m_pView->UpdateView();
}

/************************************************************
 *	@brief		Rotate
 *	@retval		Nothing
 ************************************************************/
void CStPlay::Rotate()
{
    // Rotate
    m_CurrentBlock.Rotate();
    // Update view
    m_pView->UpdateView();
}

/************************************************************
 *	@brief		Drop
 *	@retval		Nothing
 ************************************************************/
void CStPlay::Drop()
{
	// Drop
	m_CurrentBlock.Drop();
	// Set block to board
	SetBlockToBoard();
	// Clear complete line
	ClearCompleteLine();
	// Change block
	ChangeBlock();

	// Add score
	m_Score.AddScore(100);

	// Set high score
	if (m_HighScore.GetScore() < m_Score.GetScore())
		m_HighScore.SetScore(m_Score.GetScore());

    // Update view
    m_pView->UpdateView();
}

/************************************************************
 *	@brief		Get board
 *	@retval		Nothing
 ************************************************************/
CStBoard* CStPlay::GetBoard()
{
	return &m_Board;
}

/************************************************************
 *	@brief		Get current block
 *	@retval		Nothing
 ************************************************************/
CStBlock* CStPlay::GetCurrentBlock()
{
	return &m_CurrentBlock;
}

/************************************************************
 *	@brief		Get next block
 *	@retval		Nothing
 ************************************************************/
CStBlock* CStPlay::GetNextBlock()
{
	return &m_NextBlock;
}

/************************************************************
 *	@brief		Get score
 *	@retval		Nothing
 ************************************************************/
CStScore* CStPlay::GetScore()
{
	return &m_Score;
}

/************************************************************
 *	@brief		Get high score
 *	@retval		Nothing
 ************************************************************/
CStScore* CStPlay::GetHighScore()
{
	return &m_HighScore;
}

/************************************************************
 *	@brief		Set block to board
 *	@retval		Nothing
 ************************************************************/
void CStPlay::SetBlockToBoard()
{
	// Get block size
	int nXSize = m_CurrentBlock.GetXSize();
	int nYSize = m_CurrentBlock.GetYSize();

	// Set block to board
	for (int nBlockY = 0; nBlockY < nYSize; nBlockY++)
	{
		for (int nBlockX = 0; nBlockX < nXSize; nBlockX++)
		{
			if (m_CurrentBlock.GetBlock(nBlockX, nBlockY))
			{
				m_Board.SetValue(m_CurrentBlock.GetXPos() + nBlockX, m_CurrentBlock.GetYPos() + nBlockY, m_CurrentBlock.GetType());
			}
		}
	}
}

/************************************************************
 *	@brief		Clear complete line
 *	@retval		Nothing
 ************************************************************/
void CStPlay::ClearCompleteLine()
{
	// Get board size
	int nXSize = m_Board.GetXSize();
	int nYSize = m_Board.GetYSize();

	// Check if a line is completed on board
	for (int nBoardY = nYSize -1; nBoardY >= 0; )
	{
		// Count block in one line
		int nBlockCount = 0;
		for (int nBoardX = 0; nBoardX < nXSize; nBoardX++)
		{
			if (m_Board.GetValue(nBoardX, nBoardY))
			{
				nBlockCount++;
			}
		}

		// If block is completed
		if (nBlockCount == nXSize)
		{
			// Move blocks from complete block - 1 to first block
			for (int nMoveBoardY = nBoardY; nMoveBoardY > 0; nMoveBoardY--)
			{
				for (int nBoardX = 0; nBoardX < nXSize; nBoardX++)
				{
					int nValue = m_Board.GetValue(nBoardX, nMoveBoardY - 1);
					m_Board.SetValue(nBoardX, nMoveBoardY, nValue);
				}
			}

			// Clear first line
			for (int nBoardX = 0; nBoardX < nXSize; nBoardX++)
			{
				m_Board.SetValue(nBoardX, 0, 0);
			}

			// Add score
			m_Score.AddScore(1000);

			// Set high score
			if (m_HighScore.GetScore() < m_Score.GetScore())
				m_HighScore.SetScore(m_Score.GetScore());
		}
		// If block is not completed
		else
		{
			// Check next line
			nBoardY--;
		}
	}
}

/************************************************************
 *	@brief		Change block
 *	@retval		Nothing
 ************************************************************/
void CStPlay::ChangeBlock()
{
	// Set current block
	m_CurrentBlock = m_NextBlock;
	// Set next block
	CStBlock nextBlock(&m_Board);
	m_NextBlock = nextBlock;
}

File: src/StPlay.js
define(['StBlock', 'StBoard'], (StBlock, StBoard) => {
    return class StPlay {
        static get ST_NOTIFY_LEFT() { return "NOTIFY_LEFT"; }
        static get ST_NOTIFY_RIGHT() { return "NOTIFY_RIGHT"; }
        static get ST_NOTIFY_DOWN() { return "NOTIFY_DOWN"; }
        static get ST_NOTIFY_DROP() { return "NOTIFY_DROP"; }
        static get ST_NOTIFY_ROTATE() { return "NOTIFY_ROTATE"; }
        static get ST_NOTIFY_CLEAR() { return "NOTIFY_CLEAR"; }
        static get ST_NOTIFY_TETRISOVER() { return "NOTIFY_TETRISOVER"; }

        // Constructor
        constructor(notify) {
            // Board
            this.board = new StBoard();
            // Current block
            this.currentBlock = new StBlock();
            this.currentBlock.setXPos(Math.floor((this.board.getXSize() - this.currentBlock.getXSize()) / 2));
            this.currentBlock.setYPos(- this.currentBlock.getYSize());
            // Next block
            this.nextBlock = new StBlock();
            // Notify
            this.notify = notify;
        }

        // Get board
        getBoard() {
            return this.board;
        }

        // Get current block
        getCurrentBlock() {
            return this.currentBlock;
        }

        // Get next block
        getNextBlock() {
            return this.nextBlock;
        }

        // Move left
        moveLeft() {
            // Is block movable
            if (this.isMovable(this.currentBlock, this.currentBlock.getXPos() - 1, this.currentBlock.getYPos(), this.currentBlock.getRotation())) {
                // Set current X position to left
                this.currentBlock.setXPos(this.currentBlock.getXPos() - 1);
                // Notify
                this.notify.notify(StPlay.ST_NOTIFY_LEFT);
            }
        }

        // Move right
        moveRight() {
            // Is block movable
            if (this.isMovable(this.currentBlock, this.currentBlock.getXPos() + 1, this.currentBlock.getYPos(), this.currentBlock.getRotation())) {
                // Set current X position to right
                this.currentBlock.setXPos(this.currentBlock.getXPos() + 1);
                // Notify
                this.notify.notify(StPlay.ST_NOTIFY_RIGHT);
            }
        }

        // Move down
        moveDown() {
            // Is block movable
            if (this.isMovable(this.currentBlock, this.currentBlock.getXPos(), this.currentBlock.getYPos() + 1, this.currentBlock.getRotation())) {
                // Set current Y position to down
                this.currentBlock.setYPos(this.currentBlock.getYPos() + 1);
                // Notify
                this.notify.notify(StPlay.ST_NOTIFY_DOWN);
            }
            // Block is not movable
            else {
                // Set current block to board
                this.setCurrentBlockToBoard();
                // Change current block
                this.changeCurrentBlock();
                // Clear completed line
                let cleared_lines = this.board.clearCompleteLines();
                if (cleared_lines) {
                    // Notify
                    this.notify.notify(StPlay.ST_NOTIFY_CLEAR, { cleared_lines : cleared_lines });
                }
                // Tetris is over
                if (this.isTetrisOver()) {
                    // Notify
                    this.notify.notify(StPlay.ST_NOTIFY_TETRISOVER);
                }
            }
        }

        // Drop
        drop() {

            while (true) {
                // Block is movable
                if (this.isMovable(this.currentBlock, this.currentBlock.getXPos(), this.currentBlock.getYPos() + 1, this.currentBlock.getRotation())) {
                    // Set current Y position to down
                    this.currentBlock.setYPos(this.currentBlock.getYPos() + 1);
                }
                // Block is not movable
                else {
                    // Notify
                    this.notify.notify(StPlay.ST_NOTIFY_DROP);
                    break;
                }
            }

            // Set current block to board
            this.setCurrentBlockToBoard();
            // Change current block
            this.changeCurrentBlock();
            // Clear completed line
            let cleared_lines = this.board.clearCompleteLines();
            if (cleared_lines) {
                // Notify
                this.notify.notify(StPlay.ST_NOTIFY_CLEAR, { cleared_lines : cleared_lines });
            }
            // Tetris is over
            if (this.isTetrisOver()) {
                // Notify
                this.notify.notify(StPlay.ST_NOTIFY_TETRISOVER);
            }
        }

        // Rotate
        rotate() {
            // Is block movable
            if (this.isMovable(this.currentBlock, this.currentBlock.getXPos(), this.currentBlock.getYPos(), this.currentBlock.getRotation() + 1)) {
                // Rotate
                this.currentBlock.setRotation(this.currentBlock.getRotation() + 1);
                // Notify
                this.notify.notify(StPlay.ST_NOTIFY_ROTATE);
            }
        }

        // Is movable
        isMovable(block, xPos, yPos, rotation) {

            // Create block to check and set value to check
            let checkBlock = new StBlock(block);
            checkBlock.setXPos(xPos);
            checkBlock.setYPos(yPos);
            checkBlock.setRotation(rotation);

            // Get value of board
            for (let blockX = 0; blockX < checkBlock.getXSize(); blockX++) {
                for (let blockY = 0; blockY < checkBlock.getYSize(); blockY++) {

                    // Check if block to check is empty
                    if (checkBlock.getBlock(blockX, blockY) == 0) {
                        continue;
                    }

                    let boardX = checkBlock.getXPos() + blockX;
                    let boardY = checkBlock.getYPos() + blockY;

                    // Check X position of block to check
                    if (boardX < 0 || boardX >= this.board.getXSize()) {
                        return false;
                    }

                    // Check Y position of block to check
                    if (boardY >= this.board.getYSize()) {
                        return false;
                    }

                    // Check if block is already exist
                    if (this.board.getBlock(boardX, boardY) != 0) {
                        return false;
                    }
                }
            }

            return true;
        }

        // Set current block to board
        setCurrentBlockToBoard() {
            // Set current block to board
            for (let blockX = 0; blockX < this.currentBlock.getXSize(); blockX++) {
                for (let blockY = 0; blockY < this.currentBlock.getYSize(); blockY++) {
                    if (this.currentBlock.getBlock(blockX, blockY) != 0) {
                        this.board.setBlock(this.currentBlock.getXPos() + blockX, this.currentBlock.getYPos() + blockY, this.currentBlock.getType());
                    }
                }
            }
        }

        // Check if tetris is over
        isTetrisOver() {
            let isTetrisOver = false;
            for (let boardX = 0; boardX < this.board.getXSize(); boardX++) {
                if (this.board.getBlock(boardX, 0) != 0) {
                    isTetrisOver = true;
                    break;
                }
            }
            return isTetrisOver;
        }

        // Change current block
        changeCurrentBlock() {
            // Set next block to current block
            this.currentBlock = this.nextBlock;
            this.currentBlock.setXPos(Math.floor((this.board.getXSize() - this.currentBlock.getXSize()) / 2));
            this.currentBlock.setYPos(- this.currentBlock.getYSize());
            // Set next block
            this.nextBlock = new StBlock();
        }
    }
});